core/services/p2p/peer.go

```diff
@@ -0,0 +1,250 @@
+package p2p
+
+import (
+	"context"
+	"crypto/ed25519"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/smartcontractkit/libocr/networking/ragedisco"
+	"github.com/smartcontractkit/libocr/ragep2p"
+	ragetypes "github.com/smartcontractkit/libocr/ragep2p/types"
+
+	nettypes "github.com/smartcontractkit/libocr/networking/types"
+
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+
+	commonlogger "github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink-common/pkg/services"
+)
+
+var (
+	defaultGroupID    = [32]byte{0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01}
+	defaultStreamName = "stream"
+)
+
+type Peer interface {
+	services.Service
+	UpdatePeerAddresses(peerIDs []ragetypes.PeerID) error
+	Send(peerID ragetypes.PeerID, msg []byte) error
+	Receive() <-chan Message
+}
+
+type Message struct {
+	Sender  ragetypes.PeerID
+	Payload []byte
+}
+
+type PeerConfig struct {
+	PrivateKey ed25519.PrivateKey
+
+	// List of <ip>:<port> addresses.
+	ListenAddresses []string
+
+	// List of <host>:<port> addresses. If empty, defaults to ListenAddresses.
+	AnnounceAddresses []string
+
+	Bootstrappers []ragetypes.PeerInfo
+
+	// Every DeltaReconcile a Reconcile message is sent to every peer.
+	DeltaReconcile time.Duration
+
+	// Dial attempts will be at least DeltaDial apart.
+	DeltaDial time.Duration
+
+	DiscovererDatabase nettypes.DiscovererDatabase
+
+	IncomingMessageBufferSize int
+	OutgoingMessageBufferSize int
+	MaxMessageLenBytes        int
+	MessageRateLimiter        ragep2p.TokenBucketParams
+	BytesRateLimiter          ragep2p.TokenBucketParams
+
+	MetricsRegisterer prometheus.Registerer
+}
+
+type peer struct {
+	streams     map[ragetypes.PeerID]*ragep2p.Stream
+	cfg         PeerConfig
+	isBootstrap bool
+	host        *ragep2p.Host
+	discoverer  *ragedisco.Ragep2pDiscoverer
+	myID        ragetypes.PeerID
+	recvCh      chan Message
+
+	stopCh services.StopChan
+	wg     sync.WaitGroup
+	lggr   logger.Logger
+}
+
+func NewPeer(cfg PeerConfig, lggr logger.Logger) (Peer, error) {
```

Comment:
I'm generally not a fan of returning interfaces: it's idiomatic in Go to return concrete types (eg. structs) and accept interfaces. So in this case, it's fine (but odd) to return *peer and then expect any users to use an interface type throughout.

Arguably I'd say the same thing about the Peer interface type -- it should live in the consuming package (or in a types package). If its purpose is to define the public interface, you can do that on the struct directly.

Reply:
Agreed with returning concrete types (changing).

I'm a fan of having an explicit interface because then I can generate mocks with Mockery. Can that be done for structs too?
If not, I'm OK with moving the interface to a "types" sub-package. I understand the arguments for placing it in the consuming package but what if there are multiple consumers? They shouldn't depend on one another so there would be multiple copies of the same interface and mock.